triport/src/main/resources/ffmpeg/4.4/share/man/man1/ffmpeg.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FFMPEG 1"
.TH FFMPEG 1 " " " " " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
ffmpeg \- ffmpeg video converter
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
ffmpeg [\fIglobal_options\fR] {[\fIinput_file_options\fR] \-i \fIinput_url\fR} ... {[\fIoutput_file_options\fR] \fIoutput_url\fR} ...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBffmpeg\fR is a very fast video and audio converter that can also grab from
a live audio/video source. It can also convert between arbitrary sample
rates and resize video on the fly with a high quality polyphase filter.
.PP
\&\fBffmpeg\fR reads from an arbitrary number of input \*(L"files\*(R" (which can be regular
files, pipes, network streams, grabbing devices, etc.), specified by the
\&\f(CW\*(C`\-i\*(C'\fR option, and writes to an arbitrary number of output \*(L"files\*(R", which are
specified by a plain output url. Anything found on the command line which
cannot be interpreted as an option is considered to be an output url.
.PP
Each input or output url can, in principle, contain any number of streams of
different types (video/audio/subtitle/attachment/data). The allowed number and/or
types of streams may be limited by the container format. Selecting which
streams from which inputs will go into which output is either done automatically
or with the \f(CW\*(C`\-map\*(C'\fR option (see the Stream selection chapter).
.PP
To refer to input files in options, you must use their indices (0\-based). E.g.
the first input file is \f(CW0\fR, the second is \f(CW1\fR, etc. Similarly, streams
within a file are referred to by their indices. E.g. \f(CW\*(C`2:3\*(C'\fR refers to the
fourth stream in the third input file. Also see the Stream specifiers chapter.
.PP
As a general rule, options are applied to the next specified
file. Therefore, order is important, and you can have the same
option on the command line multiple times. Each occurrence is
then applied to the next input or output file.
Exceptions from this rule are the global options (e.g. verbosity level),
which should be specified first.
.PP
Do not mix input and output files \*(-- first specify all input files, then all
output files. Also do not mix options which belong to different files. All
options apply \s-1ONLY\s0 to the next input or output file and are reset between files.
.IP "\(bu" 4
To set the video bitrate of the output file to 64 kbit/s:
.Sp
.Vb 1
\&        ffmpeg \-i input.avi \-b:v 64k \-bufsize 64k output.avi
.Ve
.IP "\(bu" 4
To force the frame rate of the output file to 24 fps:
.Sp
.Vb 1
\&        ffmpeg \-i input.avi \-r 24 output.avi
.Ve
.IP "\(bu" 4
To force the frame rate of the input file (valid for raw formats only)
to 1 fps and the frame rate of the output file to 24 fps:
.Sp
.Vb 1
\&        ffmpeg \-r 1 \-i input.m2v \-r 24 output.avi
.Ve
.PP
The format option may be needed for raw input files.
.SH "DETAILED DESCRIPTION"
.IX Header "DETAILED DESCRIPTION"
The transcoding process in \fBffmpeg\fR for each output can be described by
the following diagram:
.PP
.Vb 10
\&         _\|_\|_\|_\|_\|_\|_              _\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_
\&        |       |            |              |
\&        | input |  demuxer   | encoded data |   decoder
\&        | file  | \-\-\-\-\-\-\-\-\-> | packets      | \-\-\-\-\-+
\&        |_\|_\|_\|_\|_\|_\|_|            |_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_|      |
\&                                                   v
\&                                               _\|_\|_\|_\|_\|_\|_\|_\|_
\&                                              |         |
\&                                              | decoded |
\&                                              | frames  |
\&                                              |_\|_\|_\|_\|_\|_\|_\|_\|_|
\&         _\|_\|_\|_\|_\|_\|_\|_             _\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_       |
\&        |        |           |              |      |
\&        | output | <\-\-\-\-\-\-\-\- | encoded data | <\-\-\-\-+
\&        | file   |   muxer   | packets      |   encoder
\&        |_\|_\|_\|_\|_\|_\|_\|_|           |_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_|
.Ve
.PP
\&\fBffmpeg\fR calls the libavformat library (containing demuxers) to read
input files and get packets containing encoded data from them. When there are
multiple input files, \fBffmpeg\fR tries to keep them synchronized by
tracking lowest timestamp on any active input stream.
.PP
Encoded packets are then passed to the decoder (unless streamcopy is selected
for the stream, see further for a description). The decoder produces
uncompressed frames (raw video/PCM audio/...) which can be processed further by
filtering (see next section). After filtering, the frames are passed to the
encoder, which encodes them and outputs encoded packets. Finally those are
passed to the muxer, which writes the encoded packets to the output file.
.SS "Filtering"
.IX Subsection "Filtering"
Before encoding, \fBffmpeg\fR can process raw audio and video frames using
filters from the libavfilter library. Several chained filters form a filter
graph. \fBffmpeg\fR distinguishes between two types of filtergraphs:
simple and complex.
.PP
\fISimple filtergraphs\fR
.IX Subsection "Simple filtergraphs"
.PP
Simple filtergraphs are those that have exactly one input and output, both of
the same type. In the above diagram they can be represented by simply inserting
an additional step between decoding and encoding:
.PP
.Vb 10
\&         _\|_\|_\|_\|_\|_\|_\|_\|_                        _\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_
\&        |         |                      |              |
\&        | decoded |                      | encoded data |
\&        | frames  |\e                   _ | packets      |
\&        |_\|_\|_\|_\|_\|_\|_\|_\|_| \e                  /||_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_|
\&                     \e   _\|_\|_\|_\|_\|_\|_\|_\|_\|_   /
\&          simple     _\e||          | /  encoder
\&          filtergraph   | filtered |/
\&                        | frames   |
\&                        |_\|_\|_\|_\|_\|_\|_\|_\|_\|_|
.Ve
.PP
Simple filtergraphs are configured with the per-stream \fB\-filter\fR option
(with \fB\-vf\fR and \fB\-af\fR aliases for video and audio respectively).
A simple filtergraph for video can look for example like this:
.PP
.Vb 4
\&         _\|_\|_\|_\|_\|_\|_        _\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_        _\|_\|_\|_\|_\|_\|_        _\|_\|_\|_\|_\|_\|_\|_
\&        |       |      |             |      |       |      |        |
\&        | input | \-\-\-> | deinterlace | \-\-\-> | scale | \-\-\-> | output |
\&        |_\|_\|_\|_\|_\|_\|_|      |_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_|      |_\|_\|_\|_\|_\|_\|_|      |_\|_\|_\|_\|_\|_\|_\|_|
.Ve
.PP
Note that some filters change frame properties but not frame contents. E.g. the
\&\f(CW\*(C`fps\*(C'\fR filter in the example above changes number of frames, but does not
touch the frame contents. Another example is the \f(CW\*(C`setpts\*(C'\fR filter, which
only sets timestamps and otherwise passes the frames unchanged.
.PP
\fIComplex filtergraphs\fR
.IX Subsection "Complex filtergraphs"
.PP
Complex filtergraphs are those which cannot be described as simply a linear
processing chain applied to one stream. This is the case, for example, when the graph has
more than one input and/or output, or when output stream type is different from
input. They can be represented with the following diagram:
.PP
.Vb 10
\&         _\|_\|_\|_\|_\|_\|_\|_\|_
\&        |         |
\&        | input 0 |\e                    _\|_\|_\|_\|_\|_\|_\|_\|_\|_
\&        |_\|_\|_\|_\|_\|_\|_\|_\|_| \e                  |          |
\&                     \e   _\|_\|_\|_\|_\|_\|_\|_\|_    /| output 0 |
\&                      \e |         |  / |_\|_\|_\|_\|_\|_\|_\|_\|_\|_|
\&         _\|_\|_\|_\|_\|_\|_\|_\|_     \e| complex | /
\&        |         |     |         |/
\&        | input 1 |\-\-\-\->| filter  |\e
\&        |_\|_\|_\|_\|_\|_\|_\|_\|_|     |         | \e   _\|_\|_\|_\|_\|_\|_\|_\|_\|_
\&                       /| graph   |  \e |          |
\&                      / |         |   \e| output 1 |
\&         _\|_\|_\|_\|_\|_\|_\|_\|_   /  |_\|_\|_\|_\|_\|_\|_\|_\|_|    |_\|_\|_\|_\|_\|_\|_\|_\|_\|_|
\&        |         | /
\&        | input 2 |/
\&        |_\|_\|_\|_\|_\|_\|_\|_\|_|
.Ve
.PP
Complex filtergraphs are configured with the \fB\-filter_complex\fR option.
Note that this option is global, since a complex filtergraph, by its nature,
cannot be unambiguously associated with a single stream or file.
.PP
The \fB\-lavfi\fR option is equivalent to \fB\-filter_complex\fR.
.PP
A trivial example of a complex filtergraph is the \f(CW\*(C`overlay\*(C'\fR filter, which
has two video inputs and one video output, containing one video overlaid on top
of the other. Its audio counterpart is the \f(CW\*(C`amix\*(C'\fR filter.
.SS "Stream copy"
.IX Subsection "Stream copy"
Stream copy is a mode selected by supplying the \f(CW\*(C`copy\*(C'\fR parameter to the
\&\fB\-codec\fR option. It makes \fBffmpeg\fR omit the decoding and encoding
step for the specified stream, so it does only demuxing and muxing. It is useful
for changing the container format or modifying container-level metadata. The
diagram above will, in this case, simplify to this:
.PP
.Vb 5
\&         _\|_\|_\|_\|_\|_\|_              _\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_            _\|_\|_\|_\|_\|_\|_\|_
\&        |       |            |              |          |        |
\&        | input |  demuxer   | encoded data |  muxer   | output |
\&        | file  | \-\-\-\-\-\-\-\-\-> | packets      | \-\-\-\-\-\-\-> | file   |
\&        |_\|_\|_\|_\|_\|_\|_|            |_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_\|_|          |_\|_\|_\|_\|_\|_\|_\|_|
.Ve
.PP
Since there is no decoding or encoding, it is very fast and there is no quality
loss. However, it might not work in some cases because of many factors. Applying
filters is obviously also impossible, since filters work on uncompressed data.
.SH "STREAM SELECTION"
.IX Header "STREAM SELECTION"
\&\fBffmpeg\fR provides the \f(CW\*(C`\-map\*(C'\fR option for manual control of stream selection in each
output file. Users can skip \f(CW\*(C`\-map\*(C'\fR and let ffmpeg perform automatic stream selection as
described below. The \f(CW\*(C`\-vn / \-an / \-sn / \-dn\*(C'\fR options can be used to skip inclusion of
video, audio, subtitle and data streams respectively, whether manually mapped or automatically
selected, except for those streams which are outputs of complex filtergraphs.
.SS "Description"
.IX Subsection "Description"
The sub-sections that follow describe the various rules that are involved in stream selection.
The examples that follow next show how these rules are applied in practice.
.PP
While every effort is made to accurately reflect the behavior of the program, FFmpeg is under
continuous development and the code may have changed since the time of this writing.
.PP
\fIAutomatic stream selection\fR
.IX Subsection "Automatic stream selection"
.PP
In the absence of any map options for a particular output file, ffmpeg inspects the output
format to check which type of streams can be included in it, viz. video, audio and/or
subtitles. For each acceptable stream type, ffmpeg will pick one stream, when available,
from among all the inputs.
.PP
It will select that stream based upon the following criteria:
.IP "\(bu" 4
for video, it is the stream with the highest resolution,
.IP "\(bu" 4
for audio, it is the stream with the most channels,
.IP "\(bu" 4
for subtitles, it is the first subtitle stream found but there's a caveat.
The output format's default subtitle encoder can be either text-based or image-based,
and only a subtitle stream of the same type will be chosen.
.PP
In the case where several streams of the same type rate equally, the stream with the lowest
index is chosen.
.PP
Data or attachment streams are not automatically selected and can only be included
using \f(CW\*(C`\-map\*(C'\fR.
.PP
\fIManual stream selection\fR
.IX Subsection "Manual stream selection"
.PP
When \f(CW\*(C`\-map\*(C'\fR is used, only user-mapped streams are included in that output file,
with one possible exception for filtergraph outputs described below.
.PP
\fIComplex filtergraphs\fR
.IX Subsection "Complex filtergraphs"
.PP
If there are any complex filtergraph output streams with unlabeled pads, they will be added
to the first output file. This will lead to a fatal error if the stream type is not supported
by the output format. In the absence of the map option, the inclusion of these streams leads
to the automatic stream selection of their types being skipped. If map options are present,
these filtergraph streams are included in addition to the mapped streams.
.PP
Complex filtergraph output streams with labeled pads must be mapped once and exactly once.
.PP
\fIStream handling\fR
.IX Subsection "Stream handling"
.PP
Stream handling is independent of stream selection, with an exception for subtitles described
below. Stream handling is set via the \f(CW\*(C`\-codec\*(C'\fR option addressed to streams within a
specific \fIoutput\fR file. In particular, codec options are applied by ffmpeg after the
stream selection process and thus do not influence the latter. If no \f(CW\*(C`\-codec\*(C'\fR option is
specified for a stream type, ffmpeg will select the default encoder registered by the output
file muxer.
.PP
An exception exists for subtitles. If a subtitle encoder is specified for an output file, the
first subtitle stream found of any type, text or image, will be included. ffmpeg does not validate
if the specified encoder can convert the selected stream or if the converted stream is acceptable
within the output format. This applies generally as well: when the user sets an encoder manually,
the stream selection process cannot check if the encoded stream can be muxed into the output file.
If it cannot, ffmpeg will abort and \fIall\fR output files will fail to be processed.
.SS "Examples"
.IX Subsection "Examples"
The following examples illustrate the behavior, quirks and limitations of ffmpeg's stream
selection methods.
.PP
They assume the following three input files.
.PP
.Vb 3
\&        input file \*(AqA.avi\*(Aq
\&              stream 0: video 640x360
\&              stream 1: audio 2 channels
\&        
\&        input file \*(AqB.mp4\*(Aq
\&              stream 0: video 1920x1080
\&              stream 1: audio 2 channels
\&              stream 2: subtitles (text)
\&              stream 3: audio 5.1 channels
\&              stream 4: subtitles (text)
\&        
\&        input file \*(AqC.mkv\*(Aq
\&              stream 0: video 1280x720
\&              stream 1: audio 2 channels
\&              stream 2: subtitles (image)
.Ve
.PP
Example: automatic stream selection
.IX Subsection "Example: automatic stream selection"
.PP
.Vb 1
\&        ffmpeg \-i A.avi \-i B.mp4 out1.mkv out2.wav \-map 1:a \-c:a copy out3.mov
.Ve
.PP
There are three output files specified, and for the first two, no \f(CW\*(C`\-map\*(C'\fR options
are set, so ffmpeg will select streams for these two files automatically.
.PP
\&\fIout1.mkv\fR is a Matroska container file and accepts video, audio and subtitle streams,
so ffmpeg will try to select one of each type.For video, it will select \f(CW\*(C`stream 0\*(C'\fR from \fIB.mp4\fR, which has the highest
resolution among all the input video streams.For audio, it will select \f(CW\*(C`stream 3\*(C'\fR from \fIB.mp4\fR, since it has the greatest
number of channels.For subtitles, it will select \f(CW\*(C`stream 2\*(C'\fR from \fIB.mp4\fR, which is the first subtitle
stream from among \fIA.avi\fR and \fIB.mp4\fR.
.PP
\&\fIout2.wav\fR accepts only audio streams, so only \f(CW\*(C`stream 3\*(C'\fR from \fIB.mp4\fR is
selected.
.PP
For \fIout3.mov\fR, since a \f(CW\*(C`\-map\*(C'\fR option is set, no automatic stream selection will
occur. The \f(CW\*(C`\-map 1:a\*(C'\fR option will select all audio streams from the second input
\&\fIB.mp4\fR. No other streams will be included in this output file.
.PP
For the first two outputs, all included streams will be transcoded. The encoders chosen will
be the default ones registered by each output format, which may not match the codec of the
selected input streams.
.PP
For the third output, codec option for audio streams has been set
to \f(CW\*(C`copy\*(C'\fR, so no decoding-filtering-encoding operations will occur, or \fIcan\fR occur.
Packets of selected streams shall be conveyed from the input file and muxed within the output
file.
.PP
Example: automatic subtitles selection
.IX Subsection "Example: automatic subtitles selection"
.PP
.Vb 1
\&        ffmpeg \-i C.mkv out1.mkv \-c:s dvdsub \-an out2.mkv
.Ve
.PP
Although \fIout1.mkv\fR is a Matroska container file which accepts subtitle streams, only a
video and audio stream shall be selected. The subtitle stream of \fIC.mkv\fR is image-based
and the default subtitle encoder of the Matroska muxer is text-based, so a transcode operation
for the subtitles is expected to fail and hence the stream isn't selected. However, in
\&\fIout2.mkv\fR, a subtitle encoder is specified in the command and so, the subtitle stream is
selected, in addition to the video stream. The presence of \f(CW\*(C`\-an\*(C'\fR disables audio stream
selection for \fIout2.mkv\fR.
.PP
Example: unlabeled filtergraph outputs
.IX Subsection "Example: unlabeled filtergraph outputs"
.PP
.Vb 1
\&        ffmpeg \-i A.avi \-i C.mkv \-i B.mp4 \-filter_complex "overlay" out1.mp4 out2.srt
.Ve
.PP
A filtergraph is setup here using the \f(CW\*(C`\-filter_complex\*(C'\fR option and consists of a single
video filter. The \f(CW\*(C`overlay\*(C'\fR filter requires exactly two video inputs, but none are
specified, so the first two available video streams are used, those of \fIA.avi\fR and
\&\fIC.mkv\fR. The output pad of the filter has no label and so is sent to the first output file
\&\fIout1.mp4\fR. Due to this, automatic selection of the video stream is skipped, which would
have selected the stream in \fIB.mp4\fR. The audio stream with most channels viz. \f(CW\*(C`stream 3\*(C'\fR
in \fIB.mp4\fR, is chosen automatically. No subtitle stream is chosen however, since the \s-1MP4\s0
format has no default subtitle encoder registered, and the user hasn't specified a subtitle encoder.
.PP
The 2nd output file, \fIout2.srt\fR, only accepts text-based subtitle streams. So, even though
the first subtitle stream available belongs to \fIC.mkv\fR, it is image-based and hence skipped.
The selected stream, \f(CW\*(C`stream 2\*(C'\fR in \fIB.mp4\fR, is the first text-based subtitle stream.
.PP
Example: labeled filtergraph outputs
.IX Subsection "Example: labeled filtergraph outputs"
.PP
.Vb 4
\&        ffmpeg \-i A.avi \-i B.mp4 \-i C.mkv \-filter_complex "[1:v]hue=s=0[outv];overlay;aresample" \e
\&               \-map \*(Aq[outv]\*(Aq \-an        out1.mp4 \e
\&                                        out2.mkv \e
\&               \-map \*(Aq[outv]\*(Aq \-map 1:a:0 out3.mkv
.Ve
.PP
The above command will fail, as the output pad labelled \f(CW\*(C`[outv]\*(C'\fR has been mapped twice.
None of the output files shall be processed.
.PP
.Vb 4
\&        ffmpeg \-i A.avi \-i B.mp4 \-i C.mkv \-filter_complex "[1:v]hue=s=0[outv];overlay;aresample" \e
\&               \-an        out1.mp4 \e
\&                          out2.mkv \e
\&               \-map 1:a:0 out3.mkv
.Ve
.PP
This command above will also fail as the hue filter output has a label, \f(CW\*(C`[outv]\*(C'\fR,
and hasn't been mapped anywhere.
.PP
The command should be modified as follows,
.PP
.Vb 4
\&        ffmpeg \-i A.avi \-i B.mp4 \-i C.mkv \-filter_complex "[1:v]hue=s=0,split=2[outv1][outv2];overlay;aresample" \e
\&                \-map \*(Aq[outv1]\*(Aq \-an        out1.mp4 \e
\&                                          out2.mkv \e
\&                \-map \*(Aq[outv2]\*(Aq \-map 1:a:0 out3.mkv
.Ve
.PP
The video stream from \fIB.mp4\fR is sent to the hue filter, whose output is cloned once using
the split filter, and both outputs labelled. Then a copy each is mapped to the first and third
output files.
.PP
The overlay filter, requiring two video inputs, uses the first two unused video streams. Those
are the streams from \fIA.avi\fR and \fIC.mkv\fR. The overlay output isn't labelled, so it is
sent to the first output file \fIout1.mp4\fR, regardless of the presence of the \f(CW\*(C`\-map\*(C'\fR option.
.PP
The aresample filter is sent the first unused audio stream, that of \fIA.avi\fR. Since this filter
output is also unlabelled, it too is mapped to the first output file. The presence of \f(CW\*(C`\-an\*(C'\fR
only suppresses automatic or manual stream selection of audio streams, not outputs sent from
filtergraphs. Both these mapped streams shall be ordered before the mapped stream in \fIout1.mp4\fR.
.PP
The video, audio and subtitle streams mapped to \f(CW\*(C`out2.mkv\*(C'\fR are entirely determined by
automatic stream selection.
.PP
\&\fIout3.mkv\fR consists of the cloned video output from the hue filter and the first audio
stream from \fIB.mp4\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
All the numerical options, if not specified otherwise, accept a string
representing a number as input, which may be followed by one of the \s-1SI\s0
unit prefixes, for example: 'K', 'M', or 'G'.
.PP
If 'i' is appended to the \s-1SI\s0 unit prefix, the complete prefix will be
interpreted as a unit prefix for binary multiples, which are based on
powers of 1024 instead of powers of 1000. Appending 'B' to the \s-1SI\s0 unit
prefix multiplies the value by 8. This allows using, for example:
\&'\s-1KB\s0', 'MiB', 'G' and 'B' as number suffixes.
.PP
Options which do not take arguments are boolean options, and set the
corresponding value to true. They can be set to false by prefixing
the option name with \*(L"no\*(R". For example using \*(L"\-nofoo\*(R"
will set the boolean option with name \*(L"foo\*(R" to false.
.SS "Stream specifiers"
.IX Subsection "Stream specifiers"
Some options are applied per-stream, e.g. bitrate or codec. Stream specifiers
are used to precisely specify which stream(s) a given option belongs to.
.PP
A stream specifier is a string generally appended to the option name and
separated from it by a colon. E.g. \f(CW\*(C`\-codec:a:1 ac3\*(C'\fR contains the
\&\f(CW\*(C`a:1\*(C'\fR stream specifier, which matches the second audio stream. Therefore, it
would select the ac3 codec for the second audio stream.
.PP
A stream specifier can match several streams, so that the option is applied to all
of them. E.g. the stream specifier in \f(CW\*(C`\-b:a 128k\*(C'\fR matches all audio
streams.
.PP
An empty stream specifier matches all streams. For example, \f(CW\*(C`\-codec copy\*(C'\fR
or \f(CW\*(C`\-codec: copy\*(C'\fR would copy all the streams without reencoding.
.PP
Possible forms of stream specifiers are:
.IP "\fIstream_index\fR" 4
.IX Item "stream_index"
Matches the stream with this index. E.g. \f(CW\*(C`\-threads:1 4\*(C'\fR would set the
thread count for the second stream to 4. If \fIstream_index\fR is used as an
additional stream specifier (see below), then it selects stream number
\&\fIstream_index\fR from the matching streams. Stream numbering is based on the
order of the streams as detected by libavformat except when a program \s-1ID\s0 is
also specified. In this case it is based on the ordering of the streams in the
program.
.IP "\fIstream_type\fR\fB[:\fR\fIadditional_stream_specifier\fR\fB]\fR" 4
.IX Item "stream_type[:additional_stream_specifier]"
\&\fIstream_type\fR is one of following: 'v' or 'V' for video, 'a' for audio, 's'
for subtitle, 'd' for data, and 't' for attachments. 'v' matches all video
streams, 'V' only matches video streams which are not attached pictures, video
thumbnails or cover arts. If \fIadditional_stream_specifier\fR is used, then
it matches streams which both have this type and match the
\&\fIadditional_stream_specifier\fR. Otherwise, it matches all streams of the
specified type.

triport/src/main/resources/ffmpeg/4.4/.brew/ffmpeg.rb
class Ffmpeg < Formula
  desc "Play, record, convert, and stream audio and video"
  homepage "https://ffmpeg.org/"
  url "https://ffmpeg.org/releases/ffmpeg-4.4.tar.xz"
  sha256 "06b10a183ce5371f915c6bb15b7b1fffbe046e8275099c96affc29e17645d909"
  # None of these parts are used by default, you have to explicitly pass `--enable-gpl`
  # to configure to activate them. In this case, FFmpeg's license changes to GPL v2+.
  license "GPL-2.0-or-later"
  head "https://github.com/FFmpeg/FFmpeg.git"

  livecheck do
    url "https://ffmpeg.org/download.html"
    regex(/href=.*?ffmpeg[._-]v?(\d+(?:\.\d+)+)\.t/i)
  end

  depends_on "nasm" => :build
  depends_on "pkg-config" => :build
  depends_on "aom"
  depends_on "dav1d"
  depends_on "fontconfig"
  depends_on "freetype"
  depends_on "frei0r"
  depends_on "gnutls"
  depends_on "lame"
  depends_on "libass"
  depends_on "libbluray"
  depends_on "libsoxr"
  depends_on "libvidstab"
  depends_on "libvorbis"
  depends_on "libvpx"
  depends_on "opencore-amr"
  depends_on "openjpeg"
  depends_on "opus"
  depends_on "rav1e"
  depends_on "rubberband"
  depends_on "sdl2"
  depends_on "snappy"
  depends_on "speex"
  depends_on "srt"
  depends_on "tesseract"
  depends_on "theora"
  depends_on "webp"
  depends_on "x264"
  depends_on "x265"
  depends_on "xvid"
  depends_on "xz"
  depends_on "zeromq"
  depends_on "zimg"

  uses_from_macos "bzip2"
  uses_from_macos "libxml2"
  uses_from_macos "zlib"

  on_linux do
    depends_on "libxv"
  end

  def install
    args = %W[
      --prefix=#{prefix}
      --enable-shared
      --enable-pthreads
      --enable-version3
      --enable-avresample
      --cc=#{ENV.cc}
      --host-cflags=#{ENV.cflags}
      --host-ldflags=#{ENV.ldflags}
      --enable-ffplay
      --enable-gnutls
      --enable-gpl
      --enable-libaom
      --enable-libbluray
      --enable-libdav1d
      --enable-libmp3lame
      --enable-libopus
      --enable-librav1e
      --enable-librubberband
      --enable-libsnappy
      --enable-libsrt
      --enable-libtesseract
      --enable-libtheora
      --enable-libvidstab
      --enable-libvorbis
      --enable-libvpx
      --enable-libwebp
      --enable-libx264
      --enable-libx265
      --enable-libxml2
      --enable-libxvid
      --enable-lzma
      --enable-libfontconfig
      --enable-libfreetype
      --enable-frei0r
      --enable-libass
      --enable-libopencore-amrnb
      --enable-libopencore-amrwb
      --enable-libopenjpeg
      --enable-libspeex
      --enable-libsoxr
      --enable-libzmq
      --enable-libzimg
      --disable-libjack
      --disable-indev=jack
    ]

    on_macos do
      # Needs corefoundation, coremedia, corevideo
      args << "--enable-videotoolbox"
    end

    system "./configure", *args
    system "make", "install"

    # Build and install additional FFmpeg tools
    system "make", "alltools"
    bin.install Dir["tools/*"].select { |f| File.executable? f }

    # Fix for Non-executables that were installed to bin/
    mv bin/"python", pkgshare/"python", force: true
  end

  test do
    # Create an example mp4 file
    mp4out = testpath/"video.mp4"
    system bin/"ffmpeg", "-filter_complex", "testsrc=rate=1:duration=1", mp4out
    assert_predicate mp4out, :exist?
  end
end
